### PR TITLE
[consensus] Restart safety for consensus integrity hash

### PR DESCRIPTION
**Pull request by andll**
Consensus integrity hash calculation can diverge after restart because consensus hash calculation did not handle a case when executor can replay transactions.

This can happen because executor replay transactions on the block-to-block basis, so after restart some transactions might be resent to consensus handler.

The fix simply ignores transactions with index less than last seen execution index.

### CODE REVIEW

**Review by mwtian — APPROVED** [review]

File: crates/sui-core/src/consensus_handler.rs [hunk 1]
```diff
@@ -62,6 +67,15 @@ impl ExecutionState for ConsensusHandler {
...
         let consensus_index = self.update_hash(consensus_index, &serialized_transaction);
+        let consensus_index = if let Some(consensus_index) = consensus_index {
+            consensus_index
+        } else {
+            debug!(
+                "Ignore consensus transaction at index {:?} as it appear to be already processed",
+                consensus_index
+            );
+            return;
+        };
```
Comment (akichidis) [inline crates/sui-core/src/consensus_handler.rs:78]: This check here seems great and very fast way to check an already execution. My only concern is that we now have 2 places where we check the same thing (have we executed this transaction?) - the second being here https://github.com/MystenLabs/sui/blob/63b92baaa3fd94695159edf497fa46e648c02a75/crates/sui-core/src/authority.rs#L2227

Do we need now the second check? If yes, then do you have any insight on what are the criteria for those 2 checks to diverge?

File: crates/sui-core/src/consensus_handler.rs [hunk 2]
```diff
@@ -16,37 +16,42 @@ use tracing::{debug, instrument, warn};
...
             sender,
-            hash,
+            last_seen,
         }
     }
 
-    fn update_hash(&self, index: ExecutionIndices, v: &[u8]) -> ExecutionIndicesWithHash {
-        let mut hash_guard = self
-            .hash
+    fn update_hash(&self, index: ExecutionIndices, v: &[u8]) -> Option<ExecutionIndicesWithHash> {
```
Comment (akichidis) [inline crates/sui-core/src/consensus_handler.rs:32]: I believe this method now bears quite lots of critical logic. Could you please add a test for it?

**Review by akichidis — APPROVED** [review]
Thanks @andll for this! Left a couple of comments.

File: crates/sui-core/src/consensus_handler.rs [hunk 3]
```diff
@@ -16,37 +16,42 @@ use tracing::{debug, instrument, warn};
...
+        let mut last_seen_guard = self
+            .last_seen
             .try_lock()
             .expect("Should not have contention on ExecutionState::update_hash");
+        if last_seen_guard.index <= index {
+            return None;
```
Comment (laura-makdah) [inline crates/sui-core/src/consensus_handler.rs:38]: Do we not expect the last seen index to be less than the new index, in other words aren't indexes expected to be strictly increasing?

Comment (andll) [reply crates/sui-core/src/consensus_handler.rs:41]: Yes indeed, good eye @laura-makdah !

File: crates/sui-core/src/consensus_handler.rs (hunk 2 above)
Comment (andll) [reply crates/sui-core/src/consensus_handler.rs]: Added test

File: crates/sui-core/src/consensus_handler.rs (hunk 3 above)
Comment (akichidis) [reply crates/sui-core/src/consensus_handler.rs:41]: Good eye indeed @laura-makdah , totally missed this in review!

File: crates/sui-core/src/consensus_handler.rs (hunk 2 above)
Comment (akichidis) [reply crates/sui-core/src/consensus_handler.rs]: Awesome thanks 👍 !

**Pull request merged**